### PR TITLE
[Merged by Bors] - refactor compute_hw with using less memory and less time

### PR DESCRIPTION
**Pull request by kranfix**
# What does this PR do?

Enhance the algorith used by `compute_hw` in `fluvio-spu`.

# Motivation

The code was difficult to read and also was using to much unnecessary memory with too many nested loops.

### CODE REVIEW

**Comment by EstebanBorai** [comment]
@kranfix great job!

File: crates/fluvio-spu/src/replication/leader/replica_state.rs [hunk 1]
```diff
@@ -387,52 +387,26 @@ fn compute_hw(
...
-            } else {
-                None
-            }
-        })
-        .collect();
-
-    hw_list.sort_unstable();
-    hw_list.pop()
+    // sort with O(n*log(n)) time without extra memory
+    let qualified_leos = BinaryHeap::from(qualified_leos);
```
Comment (sehz) [inline crates/fluvio-spu/src/replication/leader/replica_state.rs:405]: can you rename this variable to more precise and not duplicate prev name?

**Review by sehz — CHANGES_REQUESTED** [review]
Nice optimization.  I didn't realize BinaryHeap could be used instead.
One minor nit

**Comment by kranfix** [comment]
Merged with master for pass some checks and optimized a little more the code.

Having `n=number of qualified leos` and `m=min_lrs=valid min_replica` and `m<=n`

Originally:
- time: `O(n^2)`
- space: `O(n)`

First version of the refactor:
- time: `O(n*log(n))`
- space: `O(n)`

Last version of the refactor:
- time: `O(n*log(m))`
- space: `O(m)`

**Pull request closed without merging**